Update our main branch with beta changes

pelted opened the pull request:
Combines proc lookup and Saml::XML extraction changes into one PR.

These are both PRed upstream as well

mjobin-mdsol committed:
Set minimum test coverage (#207)

* Set minimum test coverage to a very high value for testing

* Update minimum coverage to actual current value

Zogoo committed:
[fix] Flaky test (#220)

* Decode AuthnRequest params to XML format before pass to mock

* Drop test case check

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

atish23 committed:
Update saml_idp.gemspec (#215)

Co-authored-by: Jon Phenow <jon@jphenow.com>

taketo1113 committed:
Add dependency of ostruct gem, to fix warning of ostruct was loaded (#217)

Co-authored-by: Jon Phenow <jon@jphenow.com>

aleksandr-obukhov committed:
Check if Rails is fully initialized (#216)

Zogoo committed:
Add new versions for CI (#223)

* Add new versions for CI

* Add ruby 3.1 for dev env

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

Zogoo committed:
[feat] External Signature param for Redirect Binding (#203)

* Squash commits for saml_idp gem

* Add explanation for external attributes of decode_request method

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

Zogoo committed:
[improvements] Return error to application and support SLO request validation (#224)

* Squash commits for saml_idp gem

* [feat] Allow SP config force signature validation (#16)

* Allow SP config force signature validation

* Allow SP config force signature validation

Tested with Slack with Authn request signature option
---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Don’t ignore certificates without usage (#17)

I have tested with live SAML SP apps and it works fine

* Unspecified certifciate from SP metadata

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* wip add error collector

* Fix type and rewrite request with proper validation test cases

* Try with proper way to update helper method (#19)

* Set minimum test coverage (#207)

* Set minimum test coverage to a very high value for testing

* Update minimum coverage to actual current value

* Try with proper way to update helper method

* Correctly decode and mock with correct REXML class

* Drop the min coverage

---------

Co-authored-by: Mathieu Jobin <majobin@mdsol.com>
Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Lead error render decision to gem user

* Validate the certificate's existence before verifying the signature.

* [feat] Collect request validation errors (#18)

* wip add error collector

* Fix type and rewrite request with proper validation test cases

* Lead error render decision to gem user

* Validate the certificate's existence before verifying the signature.

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Support lowercase percent-encoded sequences for URL encoding (#20)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Remove duplications

* Pre-conditions need to be defined in before section

* Le's not test logger in here

* Let's not break anything for now

* Rename correctly

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>
Co-authored-by: Mathieu Jobin <majobin@mdsol.com>

pelted committed:
Alternative to support multiple x509 Certificates via procs (#211)

* Adds support for multiple multiple x509 certificates, secret keys, and passwords by providing procs in the idp configuration.

* Call the proc in the tests

* Add documentation in the form of a comment in the README.md

* fix extra space

* remove additional change

* Fix metadata x509 certificate

Zogoo committed:
[fix] Accept reference id for SAML response (#226)

* Squash commits for saml_idp gem

* [feat] Allow SP config force signature validation (#16)

* Allow SP config force signature validation

* Allow SP config force signature validation

Tested with Slack with Authn request signature option
---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Don’t ignore certificates without usage (#17)

I have tested with live SAML SP apps and it works fine

* Unspecified certifciate from SP metadata

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Try with proper way to update helper method (#19)

* Set minimum test coverage (#207)

* Set minimum test coverage to a very high value for testing

* Update minimum coverage to actual current value

* Try with proper way to update helper method

* Correctly decode and mock with correct REXML class

* Drop the min coverage

---------

Co-authored-by: Mathieu Jobin <majobin@mdsol.com>
Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Collect request validation errors (#18)

* wip add error collector

* Fix type and rewrite request with proper validation test cases

* Lead error render decision to gem user

* Validate the certificate's existence before verifying the signature.

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Support lowercase percent-encoded sequences for URL encoding (#20)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Pass ref id as Session Index

* Official Rails 8 is not released yet to RubyGem until that let's stick official older version

* [fix] Gem CI updates for latest versions (#22)

* Remove duplications

* Pre-conditions need to be defined in before section

* Le's not test logger in here

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [fix] Allow IdP set reference ID for SAML response (#21)

* Pass ref id as Session Index

* Official Rails 8 is not released yet to RubyGem until that let's stick official older version

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Fixes for ORIGIN gem

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>
Co-authored-by: Mathieu Jobin <majobin@mdsol.com>

Zogoo committed:
Switchable assertion signature flag (#228)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

Zogoo committed:
Use concurrent ruby fixed version for test (#230)

* Squash commits for saml_idp gem

* [feat] Allow SP config force signature validation (#16)

* Allow SP config force signature validation

* Allow SP config force signature validation

Tested with Slack with Authn request signature option
---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Don’t ignore certificates without usage (#17)

I have tested with live SAML SP apps and it works fine

* Unspecified certifciate from SP metadata

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Try with proper way to update helper method (#19)

* Set minimum test coverage (#207)

* Set minimum test coverage to a very high value for testing

* Update minimum coverage to actual current value

* Try with proper way to update helper method

* Correctly decode and mock with correct REXML class

* Drop the min coverage

---------

Co-authored-by: Mathieu Jobin <majobin@mdsol.com>
Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Collect request validation errors (#18)

* wip add error collector

* Fix type and rewrite request with proper validation test cases

* Lead error render decision to gem user

* Validate the certificate's existence before verifying the signature.

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Support lowercase percent-encoded sequences for URL encoding (#20)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [fix] Gem CI updates for latest versions (#22)

* Remove duplications

* Pre-conditions need to be defined in before section

* Le's not test logger in here

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [fix] Allow IdP set reference ID for SAML response (#21)

* Pass ref id as Session Index

* Official Rails 8 is not released yet to RubyGem until that let's stick official older version

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Support rails 8 for dev env (#23)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Assertion flag should able switchable by application (#24)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* concurrent-ruby v1.3.5 has removed the dependency on logger

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>
Co-authored-by: Mathieu Jobin <majobin@mdsol.com>

taketo1113 committed:
Add Ruby 3.4 and Rails 8.0/7.2 to CI matrix (#236)

jphenow committed:
v1 (#239)

taketo1113 committed:
CI: Follow up to add Rails 8.0 to CI matrix (#240)

Zogoo committed:
[fix] Configure a certificate and private key for each response (#227)

* Squash commits for saml_idp gem

* [feat] Allow SP config force signature validation (#16)

* Allow SP config force signature validation

* Allow SP config force signature validation

Tested with Slack with Authn request signature option
---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Don’t ignore certificates without usage (#17)

I have tested with live SAML SP apps and it works fine

* Unspecified certifciate from SP metadata

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Try with proper way to update helper method (#19)

* Set minimum test coverage (#207)

* Set minimum test coverage to a very high value for testing

* Update minimum coverage to actual current value

* Try with proper way to update helper method

* Correctly decode and mock with correct REXML class

* Drop the min coverage

---------

Co-authored-by: Mathieu Jobin <majobin@mdsol.com>
Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [feat] Collect request validation errors (#18)

* wip add error collector

* Fix type and rewrite request with proper validation test cases

* Lead error render decision to gem user

* Validate the certificate's existence before verifying the signature.

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Support lowercase percent-encoded sequences for URL encoding (#20)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [fix] Gem CI updates for latest versions (#22)

* Remove duplications

* Pre-conditions need to be defined in before section

* Le's not test logger in here

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* [fix] Allow IdP set reference ID for SAML response (#21)

* Pass ref id as Session Index

* Official Rails 8 is not released yet to RubyGem until that let's stick official older version

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Support rails 8 for dev env (#23)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* Signable logic with given certificate information

* Update unit test with new test certificate

* Assertion builder with certificate attribute

* Response builder with ceritificate

* Use directly provided cert and pv key

* Remove config dependency from low layer logics

* Use correct attribute name

* Remove config dependency from low level logics

* Remove config dependency from low level logics and fix test

* Revert Proc approach

* Assertion flag should able switchable by application (#24)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* concurrent-ruby v1.3.5 has removed the dependency on logger (#27)

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>

* MetadataBuilder uses custom configurator (#25)

Co-authored-by: Andrea Lorenzetti <64900248+andnoz@users.noreply.github.com>

---------

Co-authored-by: zogoo <ch.tsogbadrakh@gmail.com>
Co-authored-by: Mathieu Jobin <majobin@mdsol.com>
Co-authored-by: Massimo Zappino <99500013+mzappino-noz@users.noreply.github.com>
Co-authored-by: Andrea Lorenzetti <64900248+andnoz@users.noreply.github.com>

taketo1113 committed:
CI: Add Ruby 4.0 and Rails 8.1 to CI Matrix (#243)

Fix a load error for the benchmark gem on Rails 6.1 with Ruby 4.0
Remove Ruby 3.2 from CI Matrix following rails-dev (Rails 8.2) support drop
Bump actions/checkout from 4 to 6

tyage committed:
Fix CSRF token leakage in SAML POST binding template (#241)

Co-authored-by: zogoo <chtsogbadrakh@gmail.com>

pelted committed:
Basic patch of our assertion extension feature for upstream introduction

- Introduced `assertion_extension` attribute in `AssertionBuilder`, `SamlResponse`, and `Controller`.
- Updated initialization methods to accept `assertion_extension` as an option.
- Enhanced the `build` methods to utilize `assertion_extension` for customizing subject confirmation data and authentication context.
- Ensured backward compatibility by maintaining existing functionality when no extension is provided.

pelted committed:
Add documentation for AssertionExtension class

- Enhanced comments in the AssertionExtension class to clarify its purpose and usage.
- Provided details on the expected implementation of subclasses and alignment with SAML 2.0 specifications.
- Added references to external documentation for further specification analysis.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/kolide/saml_idp?pullRequest=7) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/kolide/saml_idp?pullRequest=7) before we can accept your contribution.<br/>**1** out of **8** committers have signed the CLA.<br/><br/>:white_check_mark: pelted<br/>:x: mjobin-mdsol<br/>:x: Zogoo<br/>:x: atish23<br/>:x: aleksandr-obukhov<br/>:x: jphenow<br/>:x: taketo1113<br/>:x: tyage<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/kolide/saml_idp?pullRequest=7) it.</sub>

pelted committed:
Make SubjectConfirmationData extension additive instead of replacing

The SubjectConfirmationData extension point previously replaced the
entire element, requiring the extension to reimplement standard
attributes (Recipient, InResponseTo, NotOnOrAfter). This was
inconsistent with the AuthnContextDecl extension point which is
additive.

Now both extension points behave the same way: standard SAML elements
are always emitted, and the extension adds custom content inside them.
The SubjectConfirmationData element is always rendered with its standard
attributes, and the extension receives the builder to add child elements.
